File: cpp/july/day_3_Prison Cells After N Days.cpp
class Solution {
public:
    
    vector<int> get_next(vector<int> &cells){
        vector<int> temp(8);
        temp[0] = 0;
        temp[7] = 0;
        for(int i = 1; i < 7; i++){
            temp[i] = cells[i-1] == cells[i+1] ? 1 : 0;
        }
        return temp;
    }
    
    vector<int> prisonAfterNDays(vector<int>& cells, int N) {
        map<vector<int>, vector<int>> mp;
        vector<int> temp(8, 0);
        int i, j;
        if(cells[7] || cells[0]){
            temp = get_next(cells);
            mp[cells] = temp;
            cells = temp;
            N--;
        }
        for(j = 0; j < N; j++){
            if(mp.find(cells) == mp.end()){
                temp = get_next(cells);
                mp[cells] = temp;
                cells = temp;
            }else{
                break;
            }
        }
        N = j != 0 ? N%j : 0;
        if(N == 0) return cells;
        for(j = 0; j < N; j++){
            cells = mp[cells];
        }
        return cells;
    }
};

File: python3/july/day_3_Prison Cells After N Days.py
class Solution:
    def prisonAfterNDays(self, cells: List[int], N: int) -> List[int]:
        def get_next(cells):
            temp = [0] * 8
            for i in range(1, 7):
                temp[i] = 1 if cells[i-1] == cells[i+1] else 0
            return temp

        mp = dict()
        if(cells[7] or cells[0]):
            temp = get_next(cells)
            mp[str(cells)] = temp
            cells = temp
            N -= 1
        j = 0
        while j < N:
            if not (str(cells) in mp):
                temp = get_next(cells)
                mp[str(cells)] = temp
                cells = temp
            else:
                break
            j += 1
        
        N = (N %j) if j != 0 else 0
        if(N == 0):
            return cells

        for j in range(0, N):
            cells = mp[str(cells)]

        return cells
